libcrux-hkdf/src/impl_hacl.rs

```diff
@@ -1,77 +1,17 @@
-    /// The result is written to `prk`.
-    ///
-    /// Note that this function panics if `salt` or `ikm` is longer than  (2**32 - 1) bytes.
-    fn extract(prk: &mut [u8; HASH_LEN], salt: &[u8], ikm: &[u8]) -> Result<(), Error>;
-
-    /// HKDF expand using the pre-key material `prk` and `info`. The output length
-    /// is defined through the type of the `okm` parameter, that the output is written to.
-    ///
-    /// Returns nothing on success.
-    /// Returns [`Error::OkmTooLarge`] if the requested `okm_len` is too large.
-    /// Returns [`Error::ArgumentsTooLarge`] if one of `salt` or `ikm` are longer than, or
-    /// `OKM_LEN` is larger than (2**32 - 1) bytes.
-    fn expand<const OKM_LEN: usize>(
-        okm: &mut [u8; OKM_LEN],
-        prk: &[u8],
-        info: &[u8],
-    ) -> Result<(), Error>;
-
-    /// HKDF expand using the pre-key material `prk` and `info`. The output length
-    /// is defined by the parameter `okm_len`.
-    ///
-    /// Returns the key material in an array of length `okm_len` on success.
-    /// Returns [`Error::OkmTooLarge`] if the requested `okm_len` is too large.
-    /// Returns [`Error::ArgumentsTooLarge`] if `salt` or `ikm` is longer than
-    /// (2**32 - 1) bytes.
-    fn expand_vec(prk: &[u8], info: &[u8], okm_len: usize) -> Result<Vec<u8>, Error>;
-
-    /// HKDF using the `salt`, input key material `ikm`, `info`. The output length
-    /// is defined through the result type.
-    /// Calls `extract` and `expand` with the given input.
-    ///
-    /// Returns the key material in an array of length `okm_len`.
-    /// Note that this function panics if `salt` or `ikm` is longer than  (2**32 - 1) bytes.
-    fn hkdf<const OKM_LEN: usize>(
-        okm: &mut [u8; OKM_LEN],
-        salt: &[u8],
-        ikm: &[u8],
-        info: &[u8],
-    ) -> Result<(), Error> {
-        let mut prk = [0u8; HASH_LEN];
-        Self::extract(&mut prk, salt, ikm)?;
-        Self::expand(okm, &prk, info)
-    }
-
-    /// HKDF using the `salt`, input key material `ikm`, `info`. The output length
-    /// is defined by the parameter `okm_len`.
-    /// Calls `extract` and `expand` with the given input.
-    ///
-    /// Returns the key material in an array of length `okm_len`.
-    /// Note that this function panics if `salt` or `ikm` is longer than  (2**32 - 1) bytes.
-    fn hkdf_vec(salt: &[u8], ikm: &[u8], info: &[u8], okm_len: usize) -> Result<Vec<u8>, Error> {
-        let mut prk = [0u8; HASH_LEN];
-        Self::extract(&mut prk, salt, ikm)?;
-        Self::expand_vec(&prk, info, okm_len)
-    }
-}
+use crate::{Algorithm, Error, HkdfMode};
 
 macro_rules! impl_hkdf {
-    ($sname:ident,$name:ident, $mode:path, $extract:ident, $expand:ident,$hash_len:literal) => {
-        pub struct $sname;
+    ($struct_name:ident,$name:ident, $string_name:literal, $mode:path, $extract:ident, $expand:ident,$hash_len:literal) => {
+        #[doc = "Implemantation of HKDF backed by"]
```

Comment:
```suggestion
        #[doc = "Implementation of HKDF backed by "]
```
You're combining with `$string_name`, right? I've also added a space there at the end of this first part.